Magdalena/single proposal list

MiaArnix opened the pull request.

MiaArnix committed:
feat: merging of open and closed proposal list, proposal service handling of user-selected filter

MiaArnix committed:
feat: merging of open and closed proposal list, proposal service handling of user-selected filter

MiaArnix committed:
style: removed commented out code, grouped variables

nathanosdev commented on src/frontend/src/app/app.routes.ts:
```diff
@@ -5,7 +5,7 @@ export const ROUTES: Routes = [
     path: '',
     pathMatch: 'full',
-    redirectTo: 'open',
+    redirectTo: 'proposal',
```
Maybe we can put the proposal list on this URL instead. We'll probably have `/proposals` as the base URL of the whole Angular app, so if we put `/proposal` on top of that it'll be `/proposals/proposal`.

nathanosdev commented on src/frontend/src/app/core/state/proposal/proposal.mapper.ts:
```diff
@@ -14,46 +14,52 @@ import {
-  }));
+  return apiResponseList
+    .sort(
+      (response1, response2) =>
```
If the proposals are not sorted correctly then we should fix that on the canister, not the frontend.

nathanosdev commented on src/frontend/src/app/core/state/proposal/proposal.service.ts:
```diff
@@ -37,66 +41,85 @@ export class ProposalService {
 
-  public async loadOpenProposalList(): Promise<void> {
-    let getResponse: ListProposalsResponse;
+  public async loadProposalList(state?: ReviewPeriodState): Promise<void> {
```
We can simplify a bit here, and we should be `await`ing the promises:

```typescript
if (isNil(state)) {
  return await this.loadAllProposals();
}

if ('in_progress' in state) {
  return await this.loadOpenProposals(state);
}

if ('completed' in state) {
  return await this.loadClosedProposals(state);
}
```

nathanosdev commented on src/frontend/src/app/core/state/proposal/proposal.service.ts:
```diff
@@ -37,66 +41,85 @@ export class ProposalService {
+  private async loadOpenProposals(state: ReviewPeriodState): Promise<void> {
+    if (this.isCached(this.openProposalListLastLoaded)) {
+      this.currentProposalListSubject.next(this.openProposalList);
       return;
```
The `else` is redundant when there's an early return.

nathanosdev commented on src/frontend/src/app/core/state/proposal/proposal.service.ts:
```diff
@@ -37,66 +41,85 @@ export class ProposalService {
       return;
+    } else {
+      const getResponse = await this.actorService.list_proposals({
+        state: optional(state),
```
State should always be `'in_progress'`, so does it make sense to pass it in as a parameter?

nathanosdev commented on src/frontend/src/app/core/state/proposal/proposal.service.ts:
```diff
@@ -37,66 +41,85 @@ export class ProposalService {
+        state: optional(state),
+      });
+      this.handleResponse(getResponse);
+      this.openProposalList = this.currentProposalListSubject.getValue();
```
It's not immediately obvious that `handleResponse` is setting the value of `currentProposalListSubject`, so it feels weird to call that function expecting it to set that value and then retrieve it here.

I'd prefer to either return the proposals from `handleResponse`, or just use `getResponse.ok.proposals` here.

nathanosdev commented on src/frontend/src/app/pages/proposal-list/proposal-list.component.ts:
```diff
@@ -0,0 +1,342 @@
+                Voting links
+              </app-key-col>
+              <app-value-col [attr.aria-labelledby]="'proposal-links-' + i">
+                @if (proposal.proposalLinks.length > 0) {
```
There's no need to check the length of an array before looping through it. The loop will run 0 times if there are 0 elements in the array.

nathanosdev commented on src/frontend/src/app/pages/proposal-list/proposal-list.component.ts:
```diff
@@ -0,0 +1,342 @@
+                (change)="updateFilter()"
+              />
+              <label appLabel [for]="key">
+                {{ filterOptions[key].valueOf() }}
```
I'm not really a fan of treating the enum like an object, I'd prefer to just create an object for this. I think it'll be easier for us to discuss offline.

nathanosdev commented on src/frontend/src/app/pages/proposal-list/proposal-list.component.ts:
```diff
@@ -0,0 +1,342 @@
+  }
+
+  public ngOnInit(): void {
+    this.selectedFilter$.subscribe(value => {
```
If we use `this.filterForm.valueChanges` then there won't be a need to create a new observable just for the selected filter. Also this observable should `.pipe(untilDestroy())`.

MiaArnix committed:
refactor: removed sorting of service response
refactored logic of handleResponse and renamed to getProposalList
arranged conditional with early return in proposal service
refactored proposal list handling for better change detection
changes route naming

MiaArnix committed:
fix: fixed broken test, adapted route redirection

MiaArnix committed:
refactor: added filter form typechecking
extracted ok response extraction to util function
replaced ReviewPeriodState with ProposalState in components
